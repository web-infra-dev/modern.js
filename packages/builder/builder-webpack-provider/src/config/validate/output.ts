import { sharedOutputConfigSchema, z } from '@modern-js/builder-shared';
import type { CopyPluginOptions, OutputConfig } from '../../types';

const AdditionalOptionsSchema = z.partialObj({ concurrency: z.number() });

const CopyPluginPatternsSchema = z.array(z.any());

const CopyPluginOptionsSchema: z.ZodType<CopyPluginOptions> = z.object({
  patterns: CopyPluginPatternsSchema,
  options: z.optional(AdditionalOptionsSchema),
});

export const outputConfigSchema: z.ZodType<OutputConfig> =
  sharedOutputConfigSchema
    .extend({
      copy: z.union([CopyPluginOptionsSchema, CopyPluginPatternsSchema]),
      convertToRem: z.union([z.boolean(), z.instanceof(Object)]),
      externals: z.any(),
    })
    .partial();
