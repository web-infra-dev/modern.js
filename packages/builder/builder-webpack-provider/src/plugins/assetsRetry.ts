import { BuilderPlugin } from '../types';
import { isHtmlDisabled } from './html';

export function builderPluginAssetsRetry(): BuilderPlugin {
  return {
    name: 'builder-plugin-assets-retry',
    setup(api) {
      api.modifyWebpackChain(async (chain, { CHAIN_ID, target }) => {
        const config = api.getNormalizedConfig();

        if (!config.output.assetsRetry || isHtmlDisabled(config, target)) {
          return;
        }

        const { AssetsRetryPlugin } = await import(
          '../webpackPlugins/AssetsRetryPlugin'
        );
        chain
          .plugin(CHAIN_ID.PLUGIN.ASSETS_RETRY)
          .use(AssetsRetryPlugin, [config.output.assetsRetry]);
      });
    },
  };
}
