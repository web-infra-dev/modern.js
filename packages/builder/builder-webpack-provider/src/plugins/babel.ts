import {
  getBabelConfig,
  createBabelChain,
  type BabelOptions,
} from '@modern-js/babel-preset-app';
import {
  JS_REGEX,
  TS_REGEX,
  mergeRegex,
  createVirtualModule,
  getBrowserslistWithDefault,
  type BuilderContext,
} from '@modern-js/builder-shared';

import type { WebpackChain, BuilderPlugin, NormalizedConfig } from '../types';

const enableCoreJsEntry = (config: NormalizedConfig, isServer: boolean) =>
  config.output.polyfill === 'entry' && !isServer;

export const getUseBuiltIns = (config: NormalizedConfig) => {
  const { polyfill } = config.output;
  if (polyfill === 'ua' || polyfill === 'off') {
    return false;
  }
  return polyfill;
};

export function applyScriptCondition({
  rule,
  config,
  context,
  includes,
  excludes,
}: {
  rule: WebpackChain.Rule;
  config: NormalizedConfig;
  context: BuilderContext;
  includes: (string | RegExp)[];
  excludes: (string | RegExp)[];
}) {
  // compile all folders in app directory, exclude node_modules
  rule.include.add({
    and: [context.rootPath, { not: /node_modules/ }],
  });

  [...includes, ...(config.source.include || [])].forEach(condition => {
    rule.include.add(condition);
  });

  [...excludes, ...(config.source.exclude || [])].forEach(condition => {
    rule.exclude.add(condition);
  });
}

export const builderPluginBabel = (): BuilderPlugin => ({
  name: 'builder-plugin-babel',
  setup(api) {
    api.modifyWebpackChain(
      async (
        chain,
        { CHAIN_ID, getCompiledPath, target, isProd, isServer },
      ) => {
        const { applyOptionsChain, isUseSSRBundle } = await import(
          '@modern-js/utils'
        );

        const config = api.getNormalizedConfig();
        const browserslist = await getBrowserslistWithDefault(
          api.context.rootPath,
          config,
          target,
        );

        const getBabelOptions = (
          appDirectory: string,
          config: NormalizedConfig,
        ) => {
          // 1. Get styled-components options
          const styledComponentsOptions = applyOptionsChain(
            {
              // "pure" is used to improve dead code elimination in production.
              // we don't need to enable it in development because it will slow down the build process.
              pure: isProd,
              displayName: true,
              ssr: isUseSSRBundle(config),
              transpileTemplateLiterals: true,
            },
            config.tools.styledComponents,
          );

          // 2. Create babel util function about include/exclude
          const includes: Array<string | RegExp> = [];
          const excludes: Array<string | RegExp> = [];

          const babelUtils = {
            addIncludes(items: string | RegExp | Array<string | RegExp>) {
              if (Array.isArray(items)) {
                includes.push(...items);
              } else {
                includes.push(items);
              }
            },
            addExcludes(items: string | RegExp | Array<string | RegExp>) {
              if (Array.isArray(items)) {
                excludes.push(...items);
              } else {
                excludes.push(items);
              }
            },
          };

          // 3. Compute final babel config by @modern-js/babel-preset-app
          const babelOptions: BabelOptions = {
            babelrc: false,
            configFile: false,
            compact: isProd,
            ...getBabelConfig({
              target: isServer ? 'server' : 'client',
              appDirectory,
              useLegacyDecorators: !config.output.enableLatestDecorators,
              useBuiltIns: isServer ? false : getUseBuiltIns(config),
              chain: createBabelChain(),
              styledComponents: styledComponentsOptions,
              userBabelConfig: config.tools.babel,
              userBabelConfigUtils: babelUtils,
              overrideBrowserslist: browserslist,
            }),
          };

          if (config.output.charset === 'utf8') {
            babelOptions.generatorOpts = {
              jsescOption: { minimal: true },
            };
          }

          return {
            babelOptions,
            includes,
            excludes,
          };
        };

        const { rootPath } = api.context;
        const { babelOptions, includes, excludes } = getBabelOptions(
          rootPath,
          config,
        );
        const useTsLoader = Boolean(config.tools.tsLoader);
        const rule = chain.module.rule(CHAIN_ID.RULE.JS);

        applyScriptCondition({
          rule,
          config,
          context: api.context,
          includes,
          excludes,
        });

        rule
          .test(useTsLoader ? JS_REGEX : mergeRegex(JS_REGEX, TS_REGEX))
          .use(CHAIN_ID.USE.BABEL)
          .loader(getCompiledPath('babel-loader'))
          .options(babelOptions);

        /**
         * If a script is imported with data URI, it can be compiled by babel too.
         * This is used by some frameworks to create virtual entry.
         * https://webpack.js.org/api/module-methods/#import
         * @example: import x from 'data:text/javascript,export default 1;';
         */
        if (config.source.compileJsDataURI) {
          chain.module
            .rule(CHAIN_ID.RULE.JS_DATA_URI)
            .mimetype({
              or: ['text/javascript', 'application/javascript'],
            })
            .use(CHAIN_ID.USE.BABEL)
            .loader(getCompiledPath('babel-loader'))
            .options(babelOptions);
        }

        addCoreJsEntry({ chain, config, isServer });
      },
    );
  },
});

/** Add core-js-entry to every entries. */
export function addCoreJsEntry({
  chain,
  config,
  isServer,
}: {
  chain: WebpackChain;
  config: NormalizedConfig;
  isServer: boolean;
}) {
  if (enableCoreJsEntry(config, isServer)) {
    const entryPoints = Object.keys(chain.entryPoints.entries() || {});
    const coreJsEntry = createVirtualModule('import "core-js";');

    for (const name of entryPoints) {
      chain.entry(name).prepend(coreJsEntry);
    }
  }
}
