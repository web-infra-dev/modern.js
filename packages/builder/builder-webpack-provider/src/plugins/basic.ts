import type { BuilderPlugin } from '../types';

/**
 * Provide some basic configs of webpack
 */
export const builderPluginBasic = (): BuilderPlugin => ({
  name: 'builder-plugin-basic',

  setup(api) {
    api.modifyWebpackChain(async (chain, { isProd }) => {
      // The base directory for resolving entry points and loaders from the configuration.
      chain.context(api.context.rootPath);

      chain.mode(isProd ? 'production' : 'development');

      chain.merge({
        infrastructureLogging: {
          // Using `error` level to avoid `webpack.cache.PackFileCacheStrategy` logs
          level: 'error',
        },
      });

      // if the chunk size exceeds 1MiB, we will throw a warning
      chain.performance.maxAssetSize(1024 * 1024);
      chain.performance.maxEntrypointSize(1024 * 1024);

      // This will be futureDefaults in webpack 6
      chain.module.parser.merge({
        javascript: {
          exportsPresence: 'error',
        },
      });
    });
  },
});
