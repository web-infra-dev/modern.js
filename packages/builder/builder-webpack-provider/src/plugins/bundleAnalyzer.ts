import { BuilderPlugin } from '../types';

export function builderPluginBundleAnalyzer(): BuilderPlugin {
  return {
    name: 'builder-plugin-bundle-analyzer',
    setup(api) {
      api.modifyWebpackChain(
        async (chain, { CHAIN_ID, getCompiledPath, target }) => {
          const config = api.getNormalizedConfig();
          // There are two ways to open the bundle analyzer:
          // 1. Set environment variable `BUNDLE_ANALYZE`
          // 2. Set performance.bundleAnalyze config
          if (
            !process.env.BUNDLE_ANALYZE &&
            !config.performance.bundleAnalyze
          ) {
            return;
          }
          const { default: BundleAnalyzer } = await import(
            getCompiledPath('webpack-bundle-analyzer')
          );
          chain
            .plugin(CHAIN_ID.PLUGIN.BUNDLE_ANALYZER)
            .use(BundleAnalyzer.BundleAnalyzerPlugin, [
              {
                analyzerMode: 'static',
                openAnalyzer: false,
                reportFilename: `report-${target}.html`,
                ...(config.performance.bundleAnalyze || {}),
              },
            ]);
        },
      );
    },
  };
}
