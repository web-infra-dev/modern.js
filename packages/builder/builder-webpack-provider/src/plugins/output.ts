import { CSSExtractOptions } from '../types/thirdParty/css';
import {
  DEFAULT_PORT,
  addTrailingSlash,
  getDistPath,
  getFilename,
  type BuilderContext,
} from '@modern-js/builder-shared';
import { isUseCssExtract } from './css';
import type { BuilderPlugin, NormalizedConfig } from '../types';

function getPublicPath({
  config,
  isProd,
  context,
}: {
  config: NormalizedConfig;
  isProd: boolean;
  context: BuilderContext;
}) {
  const { dev, output } = config;

  let publicPath = '/';

  if (isProd) {
    if (output.assetPrefix) {
      publicPath = output.assetPrefix;
    }
  } else if (typeof dev.assetPrefix === 'string') {
    publicPath = dev.assetPrefix;
  } else if (dev.assetPrefix === true) {
    const hostname = context.devServer?.hostname || 'localhost';
    const port = context.devServer?.port || DEFAULT_PORT;
    publicPath = `//${hostname}:${port}/`;
  }

  return addTrailingSlash(publicPath);
}

export const builderPluginOutput = (): BuilderPlugin => ({
  name: 'builder-plugin-output',

  setup(api) {
    api.modifyWebpackChain(
      async (chain, { isProd, isServer, target, CHAIN_ID }) => {
        const config = api.getNormalizedConfig();
        const jsPath = getDistPath(config.output, 'js');
        const cssPath = getDistPath(config.output, 'css');

        const publicPath = getPublicPath({
          config,
          isProd,
          context: api.context,
        });

        // js output
        const jsFilename = getFilename(config.output, 'js', isProd);
        chain.output
          .path(api.context.distPath)
          .filename(`${jsPath}/${jsFilename}`)
          .chunkFilename(`${jsPath}/async/${jsFilename}`)
          .publicPath(publicPath)
          // disable pathinfo to improve compile performance
          // the path info is useless in most cases
          // see: https://webpack.js.org/guides/build-performance/#output-without-path-info
          .pathinfo(false)
          // since webpack v5.54.0+, hashFunction supports xxhash64 as a faster algorithm
          // which will be used as default when experiments.futureDefaults is enabled.
          .hashFunction('xxhash64');

        // css output
        if (isUseCssExtract(config, target)) {
          const { default: MiniCssExtractPlugin } = await import(
            'mini-css-extract-plugin'
          );
          const { applyOptionsChain } = await import('@modern-js/utils');
          const extractPluginOptions = applyOptionsChain(
            {},
            (config.tools.cssExtract as CSSExtractOptions)?.pluginOptions || {},
          );

          const cssFilename = getFilename(config.output, 'css', isProd);

          chain
            .plugin(CHAIN_ID.PLUGIN.MINI_CSS_EXTRACT)
            .use(MiniCssExtractPlugin, [
              {
                filename: `${cssPath}/${cssFilename}`,
                chunkFilename: `${cssPath}/async/${cssFilename}`,
                ignoreOrder: true,
                ...extractPluginOptions,
              },
            ]);
        }

        // server output
        if (isServer) {
          const serverPath = getDistPath(config.output, 'server');
          const filename = `${serverPath}/[name].js`;

          chain.output
            .filename(filename)
            .chunkFilename(filename)
            .libraryTarget('commonjs2');
        }
      },
    );
  },
});
