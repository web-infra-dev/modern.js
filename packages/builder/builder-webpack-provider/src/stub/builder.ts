import type * as playwright from '@modern-js/e2e/playwright';
import { getTemplatePath } from '@modern-js/utils';
import _ from '@modern-js/utils/lodash';
import {
  createPluginStore,
  createPublicContext,
  applyDefaultBuilderOptions,
  type PluginStore,
  type BuildOptions,
  type CreateBuilderOptions,
} from '@modern-js/builder-shared';
import assert from 'assert';
import { PathLike } from 'fs';
import {
  applyBasicPlugins,
  applyDefaultPlugins,
  applyMinimalPlugins,
} from '../shared/plugin';
import { URL } from 'url';
import { Hooks } from '../core/initHooks';
import type { BuilderConfig, BuilderPlugin, Context } from '../types';
import { STUB_BUILDER_PLUGIN_BUILTIN } from './constants';
import { createStubContext } from './context';
import { matchLoader, globContentJSON, filenameToGlobExpr } from './utils';

export interface OptionsPluginsItem {
  builtin?: boolean | 'default' | 'minimal' | 'basic';
  additional?: BuilderPlugin[];
}

export interface StubBuilderOptions extends CreateBuilderOptions {
  context?: Context;
  /**
   * Setup builtin plugins and add custom plugins.
   * Automatically add builtin plugins by `process.env.STUB_BUILDER_PLUGIN_BUILTIN`.
   */
  plugins?: OptionsPluginsItem | OptionsPluginsItem[keyof OptionsPluginsItem];
  /**
   * Whether to run webpack build. By default it will be `false` and skip webpack building.
   * Set a string value to specify the `output.distPath` config.
   */
  webpack?: boolean | string;
  buildOptions?: BuildOptions;
  builderConfig?: BuilderConfig;
}

export type HookApi = {
  [key in keyof Hooks]: Parameters<Parameters<Hooks[key]['tap']>[0]>;
};

export interface ServeDestOptions {
  hangOn?: playwright.TestType<any, any> | boolean;
}

export function normalizeStubPluginOptions(
  options?: StubBuilderOptions['plugins'],
) {
  const opt: Required<OptionsPluginsItem> = {
    builtin: STUB_BUILDER_PLUGIN_BUILTIN ?? false,
    additional: [],
  };
  // normalize options
  if (Array.isArray(options)) {
    opt.additional = options;
  } else if (typeof options === 'object') {
    Object.assign(opt, options);
  } else if (typeof options === 'string' || typeof options === 'boolean') {
    opt.builtin = options;
  }
  return opt;
}

export async function applyPluginOptions(
  pluginStore: PluginStore,
  options?: StubBuilderOptions['plugins'],
) {
  const opt = normalizeStubPluginOptions(options);
  // apply plugins
  if (opt.builtin === true || opt.builtin === 'minimal') {
    pluginStore.addPlugins(await applyMinimalPlugins());
  } else if (opt.builtin === 'basic') {
    pluginStore.addPlugins(await applyBasicPlugins());
  } else if (opt.builtin === 'default') {
    pluginStore.addPlugins(await applyDefaultPlugins());
  }
  pluginStore.addPlugins(opt.additional);
}

/**
 * Create stub builder for testing.
 * Some behaviors will be different to common `createBuilder`.
 */
export async function createStubBuilder(options?: StubBuilderOptions) {
  // init primary builder.
  const builderOptions = applyDefaultBuilderOptions(
    options,
  ) as Required<StubBuilderOptions>;
  // apply webpack option.
  if (options?.webpack) {
    const distPath =
      typeof options.webpack === 'string'
        ? options.webpack
        : getTemplatePath('modern-js/stub-builder/dist');
    _.set(builderOptions.builderConfig, 'output.distPath', distPath);
  }
  // init context.
  const context = createStubContext(builderOptions);
  // merge user context.
  options?.context && _.merge(context, options.context);

  const publicContext = createPublicContext(context);
  const pluginStore = createPluginStore();

  // add builtin and custom plugins by `options.plugins`.
  await applyPluginOptions(pluginStore, options?.plugins);

  // tap on each hook and cache the args.
  const resolvedHooks: Record<string, any> = {};
  _.each(context.hooks, ({ tap }, name) => {
    tap((...args) => {
      resolvedHooks[name] = args;
    });
  });

  /**
   * Run build and caching args of all hooks.
   * It will run actual build only first time,
   * and always return cached result until {@link reset}.
   */
  const build = _.memoize(async () => {
    const { build: buildImpl, webpackBuild } = await import('../core/build');
    const executeBuild = options?.webpack ? webpackBuild : undefined;
    await buildImpl(
      { context, pluginStore, builderOptions },
      options?.buildOptions,
      executeBuild,
    );
    return { resolvedHooks: { ...resolvedHooks } };
  });

  /**
   * Unwrap args of hook.
   *
   * The `unwrap` means it will try to run {@link build} and run some asserts before return.
   */
  const unwrapHook = async <T extends keyof HookApi>(
    hook: T,
  ): Promise<HookApi[T]> => (await build()).resolvedHooks[hook];

  /** Unwrap webpack configs. */
  const unwrapWebpackConfigs = async () => {
    const [{ bundlerConfigs }] = await unwrapHook('onBeforeBuildHook');
    return bundlerConfigs;
  };

  /** Unwrap webpack config, it will ensure there's only one config object. */
  const unwrapWebpackConfig = async () => {
    const webpackConfigs = await unwrapWebpackConfigs();
    assert(webpackConfigs.length === 1);
    return webpackConfigs[0];
  };

  /** Unwrap webpack compiler instance. */
  const unwrapWebpackCompiler = async () => {
    const [{ compiler }] = await unwrapHook('onAfterCreateCompilerHooks');
    return compiler;
  };

  /** Serialize content of output files into JSON object. */
  const unwrapOutputJSON = async (
    paths: PathLike | PathLike[] = context.distPath,
    isRelative = false,
    maxSize = 4096,
  ) => {
    if (Array.isArray(paths) && isRelative) {
      throw new Error('`isRelative` is not supported for multiple paths.');
    }
    await build();
    const _paths = _(paths)
      .castArray()
      .map(filenameToGlobExpr)
      .map(String)
      .value();
    return globContentJSON(_paths, { absolute: !isRelative, maxSize });
  };

  /** Read output file content. */
  const readOutputFile = async (
    filename: string,
  ): Promise<string | undefined> => {
    const compiler = await unwrapWebpackCompiler();
    return new Promise((resolve, reject) => {
      compiler.outputFileSystem.readFile(filename, (err, data) =>
        err ? reject(err) : resolve(data?.toString('utf-8')),
      );
    });
  };

  /** Unwrap output file content, will ensure it is not empty. */
  const unwrapOutputFile = async (filename: string) => {
    await build();
    const content = await readOutputFile(filename);
    assert(content);
    return content;
  };

  /**
   * Run build and serve on `distPath`.
   * If `hangOn` is true, it will block running and let you debug the server.
   * Set `hangOn` to a {@link playwright.TestType} to automatically unset timeout of test.
   *
   * @example
   * test('should work', async () => {
   *   const builder = await createStubBuilder();
   *   const { baseurl } = builder.buildAndServe({ hangOn: test });
   *   page.goto(`${baseurl}/index.html`);
   * });
   */
  const buildAndServe = async (options?: ServeDestOptions) => {
    const [{ runStaticServer }] = await Promise.all([
      import('@modern-js/e2e'),
      build(),
    ]);
    const { port } = await runStaticServer(context.distPath);
    if (options?.hangOn) {
      // eslint-disable-next-line no-console
      console.log(
        `Successfully build, and hang on running server: http://localhost:${port}`,
      );
      typeof options.hangOn !== 'boolean' && options.hangOn.setTimeout(0);
      await new Promise(() => null);
    }
    // TODO: get real `htmlRoot` & `homeUrl` from compiler.
    const baseUrl = new URL(`http://localhost:${port}`);
    const htmlRoot = new URL('html/', baseUrl);
    const homeUrl = new URL('index/index.html', htmlRoot);
    return {
      baseUrl,
      htmlRoot,
      homeUrl,
      port,
    };
  };

  /** Match webpack plugin by constructor name. */
  const matchWebpackPlugin = async (pluginName: string) => {
    const config = await unwrapWebpackConfig();
    const result = config.plugins?.filter(
      item => item.constructor.name === pluginName,
    );
    if (Array.isArray(result)) {
      assert(result.length <= 1);
      return result[0] || null;
    } else {
      return result || null;
    }
  };

  /** Check if a file handled by specific loader. */
  const matchWebpackLoader = async (loader: string, testFile: string) =>
    matchLoader({ config: await unwrapWebpackConfig(), loader, testFile });

  /** Reset builder. @see {@link build} */
  const reset = () => {
    build.cache.clear!();
  };

  return {
    ...pluginStore,
    build,
    hooks: context.hooks,
    context,
    publicContext,
    unwrapHook,
    unwrapWebpackConfigs,
    unwrapWebpackConfig,
    unwrapWebpackCompiler,
    unwrapOutputJSON,
    unwrapOutputFile,
    readOutputFile,
    buildAndServe,
    matchWebpackPlugin,
    matchWebpackLoader,
    reset,
  };
}
