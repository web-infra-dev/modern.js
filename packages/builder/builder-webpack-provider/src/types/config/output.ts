import type {
  SharedOutputConfig,
  NormalizedSharedOutputConfig,
} from '@modern-js/builder-shared';
import type { CopyPluginOptions, WebpackConfig } from '../thirdParty';

/**
 * postcss-pxtorem options
 * https://github.com/cuth/postcss-pxtorem#options
 */
export type PxToRemOptions = {
  rootValue?: number;
  unitPrecision?: number;
  propList?: Array<string>;
  selectorBlackList?: Array<string>;
  replace?: boolean;
  mediaQuery?: boolean;
  minPixelValue?: number;
  exclude?: string | RegExp | ((filePath: string) => boolean);
};

export type RemOptions = {
  /** Whether to inject runtime code into html templates。Default: true */
  enableRuntime?: boolean;
  /** Usually, `fontSize = (clientWidth * rootFontSize) / screenWidth` */
  screenWidth?: number;
  rootFontSize?: number;
  maxRootFontSize?: number;
  /** Get clientWidth from the url query based on widthQueryKey */
  widthQueryKey?: string;
  /** The entries to ignore */
  excludeEntries?: string[];
  /** Use height to calculate rem in landscape。Default: false */
  supportLandscape?: boolean;
  /** Whether to use rootFontSize when large than maxRootFontSize（scene：pc） */
  useRootFontSizeBeyondMax?: boolean;
  /** CSS (postcss-pxtorem) option */
  pxtorem?: PxToRemOptions;
};

export type ExternalsOptions = WebpackConfig['externals'];

export type OutputConfig = SharedOutputConfig & {
  /**
   * Copies the specified file or directory to the dist directory.
   */
  copy?: CopyPluginOptions | CopyPluginOptions['patterns'];
  /**
   * Convert px to rem in CSS.
   */
  convertToRem?: boolean | RemOptions;
  /**
   * At build time, prevent some `import` dependencies from being packed into bundles in your code, and instead fetch them externally at runtime.
   * For more information, please see: [webpack Externals](https://webpack.js.org/configuration/externals/)
   */
  externals?: ExternalsOptions;
};

export type NormalizedOutputConfig = OutputConfig &
  NormalizedSharedOutputConfig;
