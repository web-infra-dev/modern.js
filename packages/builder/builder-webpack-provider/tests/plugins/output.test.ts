import { expect, describe, it } from 'vitest';
import { builderPluginOutput } from '@/plugins/output';
import { createStubBuilder } from '@/stub/builder';

describe('plugins/output', () => {
  it('should set output correctly', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginOutput()],
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should allow to custom server directory with distPath.server', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginOutput()],
      target: ['node'],
      builderConfig: {
        output: {
          distPath: {
            server: 'server',
          },
        },
      },
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });

  it('should allow to use filename.js to modify filename', async () => {
    const builder = await createStubBuilder({
      plugins: [builderPluginOutput()],
      builderConfig: {
        output: {
          filename: {
            js: 'foo.js',
          },
        },
      },
    });

    const config = await builder.unwrapWebpackConfig();
    expect(config).toMatchSnapshot();
  });
});
