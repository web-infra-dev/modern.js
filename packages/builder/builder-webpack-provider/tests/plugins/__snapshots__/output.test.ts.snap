// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugins/output > should allow to custom server directory with distPath.server 1`] = `
{
  "output": {
    "chunkFilename": "server/[name].js",
    "filename": "server/[name].js",
    "hashFunction": "xxhash64",
    "libraryTarget": "commonjs2",
    "path": "<ROOT>/dist",
    "pathinfo": false,
    "publicPath": "/",
  },
}
`;

exports[`plugins/output > should allow to set distPath.js and distPath.css to empty string 1`] = `
{
  "output": {
    "chunkFilename": "async/[name].js",
    "filename": "[name].js",
    "hashFunction": "xxhash64",
    "path": "<ROOT>/dist",
    "pathinfo": false,
    "publicPath": "/",
  },
  "plugins": [
    MiniCssExtractPlugin {
      "_sortedModulesCache": WeakMap {},
      "options": {
        "chunkFilename": "async/[name].css",
        "experimentalUseImportModule": undefined,
        "filename": "[name].css",
        "ignoreOrder": true,
        "runtime": true,
      },
      "runtimeOptions": {
        "attributes": undefined,
        "insert": undefined,
        "linkType": "text/css",
      },
    },
  ],
}
`;

exports[`plugins/output > should allow to use filename.js to modify filename 1`] = `
{
  "output": {
    "chunkFilename": "static/js/async/foo.js",
    "filename": "static/js/foo.js",
    "hashFunction": "xxhash64",
    "path": "<ROOT>/dist",
    "pathinfo": false,
    "publicPath": "/",
  },
  "plugins": [
    MiniCssExtractPlugin {
      "_sortedModulesCache": WeakMap {},
      "options": {
        "chunkFilename": "static/css/async/[name].css",
        "experimentalUseImportModule": undefined,
        "filename": "static/css/[name].css",
        "ignoreOrder": true,
        "runtime": true,
      },
      "runtimeOptions": {
        "attributes": undefined,
        "insert": undefined,
        "linkType": "text/css",
      },
    },
  ],
}
`;

exports[`plugins/output > should set output correctly 1`] = `
{
  "output": {
    "chunkFilename": "static/js/async/[name].js",
    "filename": "static/js/[name].js",
    "hashFunction": "xxhash64",
    "path": "<ROOT>/dist",
    "pathinfo": false,
    "publicPath": "/",
  },
  "plugins": [
    MiniCssExtractPlugin {
      "_sortedModulesCache": WeakMap {},
      "options": {
        "chunkFilename": "static/css/async/[name].css",
        "experimentalUseImportModule": undefined,
        "filename": "static/css/[name].css",
        "ignoreOrder": true,
        "runtime": true,
      },
      "runtimeOptions": {
        "attributes": undefined,
        "insert": undefined,
        "linkType": "text/css",
      },
    },
  ],
}
`;
